### ISSUE TITLE
Make Proof General installable with OPAM

### ISSUE THREAD

**Issue by pmetzger** [ProofGeneral/PG#369]
Given that a major user pool for Proof General is the Coq community, and Coq is typically installed with OPAM these days, it might be nice to make Proof General installable with OPAM as well. Then getting both in place would just be a question of typing `opam install coq proof-general`.

The opam configs for [tuareg](https://github.com/ocaml/tuareg/tree/master/opam) might be reasonably inspirational.

**Comment by erikmd** [comment]
Hi @pmetzger, thanks for your suggestion! Indeed `tuareg` is a good example of Emacs package for OCaml users, that is thereby distributed in both MELPA and OPAM. We could do the same for PG (and we are actually very close to publishing PG on MELPA), but I have a few reservations regarding the distribution of an OPAM package for PG:

As I had said in [this comment](https://github.com/ProofGeneral/PG/issues/215#issuecomment-348508280), using MELPA would be simpler for users than relying on OPAM, because there would be no bookkeeping required w.r.t. `opam switch`... and once PG is installed using MELPA, there would be no need to further edit the `.emacs` to manually load PG's `proof-site.el`.

Finally maintaining an OPAM package would require a frequent release schedule so that recent fixes can be made available to users, while with MELPA users can automatically update their packages on-demand to benefit from the latest version of `master`...

I Cc @Matafou and @cpitclaudel in case they have other comments on this topic.

Kind regards, Erik

**Comment by cpitclaudel** [comment]
I have no strong feelings about this.  I prefer MELPA to OPAM.  But I don't object to having PG in OPAM, as long as it's regularly updated there.
Is there a way to tell OPAM to install from the repo?

**Comment by anton-trunov** [comment]
@cpitclaudel Yes, it's easy. If I have time this weekend, I can try submitting a pull request. Basically, we would need to add
- `descr` file to PG's repo (its copies will go to OPAM as well);
- an `opam` file and a `url` file to https://github.com/coq/opam-coq-archive.

I'd suggest adding the current stable release (if there is one) into `coq-released` section and `master` and `async` branches to `coq-extra-dev` as separate packages.

**Comment by cpitclaudel** [comment]
@anton-trunov Your help would be welcome.  But I wonder: do we have a way to let users install directly from the repo (rather than adding ourselves to the Coq OPAM repo?)

It'd be ideal to say "you can use opam install github.com/ProofGeneral/PG/master` or something similar.

**Comment by anton-trunov** [comment]
> It'd be ideal to say "you can use opam install github.com/ProofGeneral/PG/master` or something similar.

@cpitclaudel I wish it could be that simple, but I guess we can't do that. The `opam` file specifies the building instructions (and other meta-information), which the `opam` utility executes. It also includes the dependencies. I can add that we have recently published a library on Coq's OPAM and I found the overall experience very nice.

btw, what minimal Coq version does PG support?

**Comment by erikmd** [comment]
@cpitclaudel 
> do we have a way to let users install directly from the repo (rather than adding ourselves to the Coq OPAM repo?)

Actually we could do both: an opam file could be part of the PG Git repo itself (with a version "4.5-git" for example or more simply "dev"), and an opam file could be submitted in the OPAM repository (after we did a tag, for example version "4.5.0"). That would enable both use cases: dev version (like MELPA) and stable version (like MELPA Stable)

This approach is followed, for example, by the MathComp team:
https://github.com/math-comp/math-comp/blob/master/mathcomp/ssreflect/opam

**Comment by erikmd** [comment]
@anton-trunov
> what minimal Coq version does PG support?

I believe it is not formally stated like this in the PG doc,
anyway according to https://proofgeneral.github.io/doc/userman/ProofGeneral_12/#Coq-Proof-General several features of PG master only work with Coq >= 8.4
(while PG async requires Coq >= 8.5, because the XML protocol is not available before 8.5 IIRC)
